src/Grpc.Net.Client/ChannelBuilder.cs

```diff
@@ -0,0 +1,157 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Net.Http;
+using Grpc.Core;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Grpc.Net.Client
+{
+    /// <summary>
+    /// A builder for configuring <see cref="GrpcChannel"/> instances.
+    /// </summary>
+    public sealed class ChannelBuilder
```

Comment:
I'm not sure if ASP.NET Core style configuration adds much. DI allows multiple lambdas to be registered which doesn't apply in this situation.

If we didn't have a builder then it would be a choice between something like:

`GrpcChannel.Create(new GrpcChannelOptions { })` vs `GrpcChannel.Create(options => { })`

The nice thing about the `ChannelBuilder.ForHttpClient(client)` is it forces the transport to be set and provides a nice helper method to do so.

The alternative is something like:

```csharp
GrpcChannel.Create(new GrpcChannelOptions
{
    TransportOptions = new HttpClientTransportOptions { HttpClient = httpClient }
});
```

Or the builder could be the argument to create and it could have some helper methods for setting some values:

```csharp
GrpcChannel.Create(builder =>
{
    builder.SetHttpClientTransport(httpClient);
    builder.SendMaxMessage = 100;
});
```

The downside is if/when we support multiple transports then people could incorrectly do something like:

```csharp
GrpcChannel.Create(builder =>
{
    // Which is used?
    builder.SetHttpClientTransport(httpClient);
    builder.SetSocketTransport(socket);
    builder.SendMaxMessage = 100;
});
```